### PR TITLE
cli: add export-login command

### LINKED ISSUE #1711: Develop enable-ci circle-ci
Estimate: 3 days

### PR DESCRIPTION
**Pull request by kyrofa**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

This PR resolves #1711 by adding the `export-login` command as well as a `--with` option to `login` to use the exported login file. This is the agreed-upon method for supporting Circle CI. See the issue as well as the [forum topic](https://forum.snapcraft.io/t/better-support-for-other-ci-systems/2968) for more details.

### CODE REVIEW

**Comment by sergiusens** [comment]
hmm I wonder why `./runtests.sh static` has been ticked

**Comment by sergiusens** [comment]
good catch! :-)

**Comment by sergiusens** [comment]
Looks like a nice cleanup. Any possibly of handling the input for user and password from `snapcraft.cli.store`? `click` should allow us to remove that crufty input mechanism

**Comment by kyrofa** [comment]
> good catch! :-)

Here you go: #1783

> Any possibly of handling the input for user and password from snapcraft.cli.store? click should allow us to remove that crufty input mechanism

I can, although this PR is already pretty big, and that change is unrelated. I'd rather it be another PR.

**Comment by DarkPaladinUHC** [comment]
Thanks a lot for working on this!

**Comment by sergiusens** [comment]
For anyone reviewing, here's a snippet on how it would work:
```
(snapcraft) ubuntu@snapcraft-dev:~/git$ snapcraft export-login --help
Usage: snapcraft export-login [OPTIONS] FILE

  Save attenuated login configuration for a store account in FILE.

  This file can then be used to login to the given account with the
  permissions specified.

  For example, to limit access to the edge channel of any snap to which the
  account has access:

      snapcraft export-login --channels=edge

  Or to limit access to only the edge channel of a single snap:

      snapcraft export-login --packages=my-snap --channels=edge

Options:
  --packages <packages>  Comma-separated list of packages to limit token
                         access
  --channels <channels>  Comma-separated list of channels to limit token
                         access
  --acls <acls>          Comma-separated list of ACLs to limit token access.
                         Valid ACLs are 'package_upload', 'package_access',
                         and 'package_manage'
  --help                 Show this message and exit.
(snapcraft) ubuntu@snapcraft-dev:~/git$ snapcraft export-login --acls package_upload my-login
Enter your Ubuntu One e-mail address and password.
If you do not have an Ubuntu One account, you can create one at https://dashboard.snapcraft.io/openid/login
Email: sergio.schvezov@canonical.com
Password: 
Second-factor auth: 272379

Login successfully exported to 'my-login'. This file can now be used with 'snapcraft login --with my-login' to login to this account with no password and have the following capabilities:

snaps:       No restriction
channels:    No restriction
permissions: ['package_upload']

This exported login is not encrypted. Do not commit it to version control!
```

File: snapcraft/cli/store.py [hunk 1]
```diff
@@ -252,18 +278,91 @@ def list_registered():
...
+@click.option('--packages', metavar='<packages>',
+              help='Comma-separated list of packages to limit token access')
+@click.option('--channels', metavar='<channels>',
+              help='Comma-separated list of channels to limit token access')
+@click.option('--acls', metavar='<acls>',
+              help='Comma-separated list of ACLs to limit token access. Valid '
+                   "ACLs are 'package_upload', 'package_access', and "
+                   "'package_manage'")
+def export_login(login_file: TextIO, packages: str, channels: str, acls: str):
+    """Save attenuated login configuration for a store account in FILE.
```
Comment (matthewpaulthomas) [inline snapcraft/cli/store.py]: Is a Snapcraft user expected to know what “attenuated” means in this context? This seems to be the remains of [removing the word “macaroon”](https://forum.snapcraft.io/t/better-support-for-other-ci-systems/2968/3) (since [an “attenuated macaroon” is apparently a thing](https://www.google.co.uk/search?q=%22attenuated+macaroon%22)).

File: snapcraft/cli/store.py [hunk 2]
```diff
@@ -252,18 +278,91 @@ def list_registered():
...
+
+    This file can then be used to login to the given account with the
```
Comment (matthewpaulthomas) [inline snapcraft/cli/store.py]: The command is `login`, but the English verb is “log in”. You could use quotes to clarify that you’re using the command name here.

File: snapcraft/cli/store.py [hunk 3]
```diff
@@ -252,18 +278,91 @@ def list_registered():
...
+    permissions specified.
+
+    For example, to limit access to the edge channel of any snap to which the
+    account has access:
```
Comment (matthewpaulthomas) [inline snapcraft/cli/store.py]: Maybe “of any snap that the account has access to:”

File: snapcraft/cli/store.py [hunk 4]
```diff
@@ -252,18 +278,91 @@ def list_registered():
...
+                           acls=acl_list,
+                           save=False):
+        sys.exit(1)
+
+    store.conf.save(config_fd=login_file)
+
+    print()
+    echo.info(
+        'Login successfully exported to {0!r}. This file can now be used with '
+        "'snapcraft login --with {0}' to login to this account with no "
```
Comment (matthewpaulthomas) [inline snapcraft/cli/store.py]: to log in to

File: snapcraft/cli/store.py [hunk 5]
```diff
@@ -252,18 +278,91 @@ def list_registered():
...
+        'password and have the following capabilities:\n'.format(
```
Comment (matthewpaulthomas) [inline snapcraft/cli/store.py]: “the following” would be shorter as “these”.

File: snapcraft/cli/store.py [hunk 6]
```diff
@@ -252,18 +278,91 @@ def list_registered():
...
+    store = storeapi.StoreClient()
+    if not snapcraft.login(store=store, config_fd=login_file):
         sys.exit(1)
 
     print()
-    echo.info('Login successful.')
+
+    if login_file:
+        echo.info(
+            'Login successful. You now have the following capabilities:\n')
```
Comment (matthewpaulthomas) [inline snapcraft/cli/store.py]: As above, “the following” would be shorter as “these”.

File: snapcraft/storeapi/__init__.py [hunk 7]
```diff
@@ -182,6 +189,7 @@ def _refresh_if_necessary(self, func, *args, **kwargs):
         try:
             return func(*args, **kwargs)
         except errors.StoreMacaroonNeedsRefreshError:
+            print('REFRESHING')
```
Comment (matthewpaulthomas) [inline snapcraft/storeapi/__init__.py]: Is this an error message, or progress text?

File: snapcraft/cli/store.py (hunk 1 above)
Comment (kyrofa) [reply snapcraft/cli/store.py]: No, I suspect they won't, this is my technical knowledge infecting the help doc :stuck_out_tongue: . What do you think this should be without getting too lengthy? "Save properly-scoped login [...]"; "Save restricted login [...]"; "Save limited login [...]" any of those sound promising?

Comment (matthewpaulthomas) [reply snapcraft/cli/store.py]: Why is it important to know that it’s attenuated/scoped/restricted? What can’t it do?

Comment (kyrofa) [reply snapcraft/cli/store.py]: "limited-access login"?

File: snapcraft/cli/store.py (hunk 3 above)
Comment (kyrofa) [reply snapcraft/cli/store.py]: I generally try to avoid ending with a preposition, although I agree that this is too wordy. Perhaps "of any snap the account can access" ?

File: snapcraft/storeapi/__init__.py (hunk 7 above)
Comment (kyrofa) [reply snapcraft/storeapi/__init__.py]: Whoops, a debug message I missed, thank you :grin: .

File: snapcraft/cli/store.py (hunk 2 above)
Comment (kyrofa) [reply snapcraft/cli/store.py]: Ah, I'm not even really talking about the command here, more the action indeed. I'll use "log in".

File: snapcraft/cli/store.py (hunk 4 above)
Comment (kyrofa) [reply snapcraft/cli/store.py]: Fixed, thank you!

File: snapcraft/cli/store.py (hunk 5 above)
Comment (kyrofa) [reply snapcraft/cli/store.py]: Agreed, done.

File: snapcraft/cli/store.py (hunk 6 above)
Comment (kyrofa) [reply snapcraft/cli/store.py]: Fixed.

File: snapcraft/cli/store.py (hunk 1 above)
Comment (kyrofa) [reply snapcraft/cli/store.py]: `export-login` exports a login that can be used to login (with no password) to the account that was exported. If the user provides no parameters, then whoever logs in with this file gains the full ability of the account that was exported (can access all their snaps, upload to them, publish to all channels, etc.). This is a lot of power to give a CI system. If I know that my CI only pushes to `edge` for snap `foo`, then I can limit that exported login to only have access to the `edge` channel of snap `foo` instead of everything by using `snapcraft export-login --packages foo -channels edge

Comment (kyrofa) [reply snapcraft/cli/store.py]: > I do think that's important to convey somehow.

But perhaps not in the title of the help text, eh?

**Comment by kyrofa** [comment]
> Please force some line breaks at 80 chars like we already do for exceptions.

We do? I don't really see that anywhere, but I see [numerous](https://github.com/snapcore/snapcraft/blob/master/snapcraft/internal/errors.py#L154) [examples](https://github.com/snapcore/snapcraft/blob/master/snapcraft/internal/errors.py#L205) that surpass 80 characters. Are we doing something magical that I'm missing?

I wish Click could help us out here, like it does with the help text. Regardless, I've added some hard wraps.

**Comment by kyrofa** [comment]
> There is an option called `packages`, isn't that supposed to be `snaps` or `snap-names`? Ask @matthewpaulthomas as I think he has been consolidating this in the docs.

Yeah I was just exposing the store API language here. @matthewpaulthomas would this be better as "snaps"?

> This table at the end
> ```
> snaps:       No restriction
> channels:    No restriction
> permissions: ['package_upload']
> ```
> 
> Without having been through the options is really confusing at first. `snaps` have `No restriction` for what?

No restriction on snaps or channels: the person who logs in with this file can utilize that ACL for any snap and any channel the account can access. I'm not sure what that really means with the `package_upload` acl, but yeah. How would you prefer this information be displayed? Or would you prefer nothing at all?

File: snapcraft/cli/store.py [hunk 8]
```diff
@@ -252,18 +278,89 @@ def list_registered():
     snapcraft.list_registered()
 
 
+@storecli.command('export-login')
+@click.argument('login_file', metavar='FILE', type=click.File('w'))
+@click.option('--packages', metavar='<packages>',
```
Comment (sergiusens) [inline snapcraft/cli/store.py]: you are still calling this packages

**Comment by sergiusens** [comment]
The store api is not the best to guide through, if so we'd have 'snapcraft push <package-file>`

This language needs to be consistent.

File: snapcraft/cli/store.py (hunk 1 above)
Comment (matthewpaulthomas) [reply snapcraft/cli/store.py]: Indeed, it might be misleading to emphasize that the login is restricted, when by default it is not.

**Comment by matthewpaulthomas** [comment]
>> There is an option called packages, isn't that supposed to be snaps or snap-names? Ask @matthewpaulthomas as I think he has been consolidating this in the docs.
>
> Yeah I was just exposing the store API language here. @matthewpaulthomas would this be better as "snaps"?

According to GitHub I’ve never changed any Snapcraft docs. Perhaps you’re thinking of @Greggless recently changing a form title, and [changing the corresponding doc](https://github.com/canonical-docs/snappy-docs/commit/b2bd02b34d121d2558a8b750c8e4a4c74f87e7fe) from “Registering a name for you snap” [sic] to “Register a snap”, because that particular form covered more than just the name.

I would use “snaps” here too, because you’re limiting access to a snap as a whole, not just its name. “packages” would be better if it was plausible that (a) the snap format would be renamed or (b) `snapcraft` would grow to cover packages that weren’t snaps but shared the same namespace. But if either of those happened, `snapcraft` probably would be renamed too, so no point in being generic.

**Comment by sergiusens** [comment]
As always, good insight, really appreciated. I might have crossed the references there

**Comment by kyrofa** [comment]
> The store api is not the best to guide through

Oh, agreed.

> I would use “snaps” here too

Excellent, done!

File: snapcraft/cli/store.py (hunk 1 above)
Comment (kyrofa) [reply snapcraft/cli/store.py]: Very good, I've removed it from the title.
